### PR TITLE
Add support for User Code Namespaces [CN-1131]

### PR DESCRIPTION
**Pull request by SeriyBg**
## User Impact

Possibility to configure User Code Namespaces via UserCodeNamespace CR

### CODE REVIEW

File: helm-charts/hazelcast-platform-operator/templates/webhook.yaml [hunk 1]
```diff
@@ -194,6 +194,29 @@ webhooks:
...
     namespaceSelector:
       {{- include "watched-namespaces.labelSelector" . | nindent 6 }}
+  - admissionReviewVersions:
+      - v1
+    clientConfig:
+      caBundle: {{ $ca.Cert | b64enc }}
+      service:
+        name: {{ $fullname }}
+        namespace: {{ .Release.Namespace }}
+        path: /validate-hazelcast-com-v1alpha1-topic
```
Comment (cheels) [inline helm-charts/hazelcast-platform-operator/templates/webhook.yaml:204]: here should be /validate-hazelcast-com-v1alpha1-usercodenamespace

File: api/v1alpha1/usercodenamespace_webhook.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *UserCodeNamespace) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	usercodenamespacelog.Info("validate update", "name", r.Name)
+	return admission.Warnings{}, ValidateUCNSpecUpdate(r)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *UserCodeNamespace) ValidateDelete() (admission.Warnings, error) {
+	usercodenamespacelog.Info("validate delete", "name", r.Name)
+	return admission.Warnings{}, nil
```
Comment (kutluhanmetin) [inline api/v1alpha1/usercodenamespace_webhook.go:39]: Just to be more consistent, we can add `ValidateUCNSpecDelete` method which returns nil or we can delete `ValidateUCNSpecUpdate` and `ValidateCreate` methods.

File: internal/controller/hazelcast/hazelcast_resources.go [hunk 3]
```diff
@@ -2550,6 +2649,10 @@ func volumes(h *hazelcastv1alpha1.Hazelcast) []v1.Volume {
 	if h.Spec.JetEngineConfiguration.IsConfigMapEnabled() {
 		vols = append(vols, configMapVolumes(jetConfigMapName, h.Spec.JetEngineConfiguration.RemoteFileConfiguration)...)
 	}
+	//
```
Comment (kutluhanmetin) [inline internal/controller/hazelcast/hazelcast_resources.go:2652]: Shall we delete them if they are not necessary?

File: internal/hazelcast-client/usercodenamespace_service.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func NewUsercodeNamespaceService(client Client) *UsercodeNamespaceService {
+	return &UsercodeNamespaceService{
+		client: client,
+	}
+}
+
+type ucnResourceType int32
+
+const (
+	class ucnResourceType = iota + 3
```
Comment (kutluhanmetin) [inline internal/hazelcast-client/usercodenamespace_service.go:26]: do we need `class` and `jar`? They are unused.

**Review by kutluhanmetin — APPROVED** [review]

**Review by hasancelik — APPROVED** [review]

File: internal/controller/hazelcast/usercodenamespace_controller.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		return errors.New("failed to get MTLS client")
+	}
+	for _, m := range client.OrderedMembers() {
+		m := m
+		g.Go(func() error {
+			host, _, err := net.SplitHostPort(m.Address.String())
+			if err != nil {
+				return err
+			}
+			bs, err := rest.NewBundleService("https://"+host+":8443", mtlsClient)
```
Comment (kutluhanmetin) [inline internal/controller/hazelcast/usercodenamespace_controller.go:166]: Shall we move 8443 to a constant in constants.go?

Comment (SeriyBg) [reply internal/controller/hazelcast/usercodenamespace_controller.go]: Make sense. We actually have a function for building `AgentUrl`, I have reused it here

File: internal/controller/hazelcast/map_controller.go [hunk 6]
```diff
@@ -218,6 +218,7 @@ func (r *MapReconciler) ReconcileMapConfig(
 			false,
 			m.Spec.Eviction.MaxSize,
 			hazelcastv1alpha1.EncodeMaxSizePolicy[m.Spec.Eviction.MaxSizePolicy],
+			defaultWanReplicationRefCodec(hz, m),
```
Comment (hasancelik) [inline internal/controller/hazelcast/map_controller.go:221]: Why did you add this method now?

File: test/e2e/usercodenamespace_test.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+
+	hazelcastcomv1alpha1 "github.com/hazelcast/hazelcast-platform-operator/api/v1alpha1"
+	"github.com/hazelcast/hazelcast-platform-operator/test"
+	hazelcastconfig "github.com/hazelcast/hazelcast-platform-operator/test/e2e/config/hazelcast"
+)
+
+var _ = Describe("Hazelcast User Code Deployment", Group("user_code_namespace"), func() {
```
Comment (hasancelik) [inline test/e2e/usercodenamespace_test.go:20]:
```suggestion
var _ = Describe("Hazelcast User Code Namespace", Group("user_code_namespace"), func() {
```

File: test/e2e/usercodenamespace_test.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+			},
+		}
+		m := hazelcastconfig.Map(ms, mapLookupKey, labels)
+		Expect(k8sClient.Create(context.Background(), m)).Should(Succeed())
+		assertMapStatus(m, hazelcastcomv1alpha1.MapSuccess)
+
+		assertMapEntryListener(h, m)
+	})
+
+	It("verify added UCN added before the HZ cluster creation is successful", Tag(Kind|Any), func() {
```
Comment (hasancelik) [inline test/e2e/usercodenamespace_test.go:167]: Can we use Hazelcast CR with executor service instead of Map CR with an entry listener?

**Review by hasancelik — CHANGES_REQUESTED** [review]

File: test/e2e/usercodenamespace_test.go (hunk 8 above)
Comment (SeriyBg) [reply test/e2e/usercodenamespace_test.go:167]: It turns out we cannot, as checking the executor service requires Java client call, so it's not possible from our operator E2E tests

File: internal/controller/hazelcast/map_controller.go (hunk 6 above)
Comment (SeriyBg) [reply internal/controller/hazelcast/map_controller.go:221]: I needed to update the coded, and the version includes this change as well

File: config/crd/bases/hazelcast.com_maps.yaml [hunk 9]
```diff
@@ -390,6 +390,10 @@ spec:
                   updated for this time are evicted automatically. It can be updated.
                 format: int32
                 type: integer
+              userCodeNamespace:
```
[the remark this answers is not in the document]
Comment (SeriyBg) [reply config/crd/bases/hazelcast.com_maps.yaml:393]: What do you mean here?

File: api/v1alpha1/hazelcast_types.go [hunk 10]
```diff
@@ -188,6 +188,10 @@ type HazelcastSpec struct {
 	// CPSubsystem is the configuration of the Hazelcast CP Subsystem.
 	// +optional
 	CPSubsystem *CPSubsystem `json:"cpSubsystem,omitempty"`
+
+	// UserCodeNamespaces provide a container for Java classpath resources, such as user code and accompanying artifacts like property files
+	// +optional
+	UserCodeNamespaces *UserCodeNamespacesConfig `json:"userCodeNamespaces,omitempty"`
```
Comment (semihbkgr) [inline api/v1alpha1/hazelcast_types.go:194]: nit: why is the name of this field plural?

Comment (SeriyBg) [reply api/v1alpha1/hazelcast_types.go:194]: That's the name of the feature: https://docs.hazelcast.com/hazelcast/5.4/clusters/user-code-namespaces

File: config/crd/bases/hazelcast.com_maps.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (hasancelik) [reply config/crd/bases/hazelcast.com_maps.yaml:393]: resolved here --> https://github.com/hazelcast/hazelcast-platform-operator/pull/1091#discussion_r1580651627

**Review by hasancelik — APPROVED** [review]

**Pull request merged**